service/frontend/nexus_http_handler.go

```diff
@@ -160,6 +156,103 @@ func (h *NexusHTTPHandler) dispatchNexusTaskByNamespaceAndTaskQueue(w http.Respo
+
+	h.serveResolvedURL(w, r, u)
+}
+
+// Handler for [nexushttp.RouteSet.DispatchNexusTaskByService].
+func (h *NexusHTTPHandler) dispatchNexusTaskByService(w http.ResponseWriter, r *http.Request) {
+	if !h.enabled() {
+		h.writeNexusFailure(w, http.StatusNotFound, &nexus.Failure{Message: "nexus endpoints disabled"})
+		return
+	}
+
+	service := parseRequestParams(commonnexus.RouteDispatchNexusTaskByService, w, r)
+
+	serviceID, err := url.PathUnescape(*service)
+	if err != nil {
+		h.logger.Error("invalid URL", tag.Error(err))
+		h.writeNexusFailure(w, http.StatusBadRequest, &nexus.Failure{Message: "invalid URL"})
+		return
+	}
+	serviceInfo, err := h.incomingServiceRegistry.Get(r.Context(), serviceID)
+	if err != nil {
+		h.logger.Error("invalid Nexus incoming service ID", tag.Error(err))
+
+		var notFound *serviceerror.NotFound
+		switch {
+		case errors.As(err, &notFound):
+			h.writeNexusFailure(w, http.StatusNotFound, &nexus.Failure{Message: err.Error()})
+		case errors.Is(err, context.DeadlineExceeded):
+			h.writeNexusFailure(w, http.StatusRequestTimeout, &nexus.Failure{Message: err.Error()})
+		default:
+			h.writeNexusFailure(w, http.StatusInternalServerError, &nexus.Failure{Message: err.Error()})
+		}
+		return
+	}
+
+	nc := h.getNexusContextFromService(serviceInfo)
+
+	r, err = h.parseTlsAndAuthInfo(r, &nc)
+	if err != nil {
+		h.logger.Error("failed to get claims", tag.Error(err))
+		h.writeNexusFailure(w, http.StatusUnauthorized, &nexus.Failure{Message: "unauthorized"})
+		return
+	}
+
+	u, err := mux.CurrentRoute(r).URL("service", *service)
+	if err != nil {
+		h.logger.Error("invalid URL", tag.Error(err))
+		h.writeNexusFailure(w, http.StatusInternalServerError, &nexus.Failure{Message: "internal error"})
+		return
+	}
+
+	h.serveResolvedURL(w, r, u)
+}
+
+func (h *NexusHTTPHandler) getBaseNexusContext(apiName string) nexusContext {
+	return nexusContext{
+		namespaceValidationInterceptor:       h.namespaceValidationInterceptor,
+		namespaceRateLimitInterceptor:        h.namespaceRateLimitInterceptor,
+		namespaceConcurrencyLimitInterceptor: h.namespaceConcurrencyLimitInterceptor,
+		rateLimitInterceptor:                 h.rateLimitInterceptor,
+		apiName:                              apiName,
+	}
+}
+
+func (h *NexusHTTPHandler) getNexusContextFromService(service *nexuspb.IncomingService) nexusContext {
+	nc := h.getBaseNexusContext(configs.DispatchNexusTaskByServiceAPIName)
+	nc.namespaceName = service.Spec.Namespace
+	nc.taskQueue = service.Spec.TaskQueue
+	nc.serviceName = service.Spec.Name
+	return nc
+}
+
+func parseRequestParams[T any](route routing.Route[T], w http.ResponseWriter, r *http.Request) *T {
+	// Limit the request body to max allowed Payload size.
+	// Content headers are transformed to Payload metadata and contribute to the Payload size as well. A separate
+	// limit is enforced on top of this in the nexusHandler.StartOperation method.
+	r.Body = http.MaxBytesReader(w, r.Body, rpc.MaxNexusAPIRequestBodyBytes)
```

Comment:
Doesn't belong in a function named `parseRequestParams`

Reply:
I see you also have the feature flag condition that can be shared.
Maybe just call this prepareRequest or split into two functions?

Reply:
And the helper as I have written it can't be a method of the handler since it is generic, so we would have to pass in a bool just to check it in the helper which I'm not a big fan of.